### PR TITLE
Add quic-client module

### PR DESCRIPTION
**Pull request by ryleung-solana**
#### Problem
See #22726  and #22728 
#### Summary of Changes
Adds a quic-client module to send transactions using quic and modifies thin-client to use it.
Fixes #

### CODE REVIEW

**Review by t-nelson — COMMENTED** [review]
This is a great start!  I think the PR is getting too large though.  Can we break out just the basic `QuicClient` that we'll build all of the other consumers on, get that fully fleshed out, test and in. Then bring in each of the implementation independently as well?  It's easier to reason about smaller changes

File: client/src/quic_client.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        _scts: &mut dyn Iterator<Item = &[u8]>,
+        _ocsp_response: &[u8],
+        _now: std::time::SystemTime,
+    ) -> Result<rustls::client::ServerCertVerified, rustls::Error> {
+        Ok(rustls::client::ServerCertVerified::assertion())
+    }
+}
+
+#[derive(Debug, Clone)]
+enum ConnectionState {
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply client/src/quic_client.rs]: Got rid of the states, and lock pressure should be much less now - not sure if the exact design was what you had in mind though.

File: client/src/thin_client.rs [hunk 2]
```diff
@@ -118,42 +120,51 @@ impl ClientOptimizer {
...
-            transactions_socket,
-            RpcClient::new_socket(rpc_addr),
-        )
+    /// and the Tpu at `quic_tpu_addr` over `transactions_socket` using QUIC.
+    /// TODO: Add UDP support if quic is set to false
+    pub fn new(
+        rpc_addr: SocketAddr,
+        quic_tpu_addr: SocketAddr,
+        transactions_socket: UdpSocket,
+        quic: bool,
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply client/src/thin_client.rs:136]: Removed most of the quic-specific references in thin-client

File: sdk/src/transport.rs [hunk 3]
```diff
@@ -10,6 +11,18 @@ pub enum TransportError {
...
+    SendDatagramError(#[from] SendDatagramError),
+
+    #[error("transport connect error: {0}")]
+    ConnectionError(#[from] ConnectionError),
+
+    #[error("transport connect error: {0}")]
+    ConnectError(#[from] ConnectError),
+
+    #[error("transport connect error: {0}")]
+    WriteError(#[from] WriteError),
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply sdk/src/transport.rs]: Made the errors more generic, and it should be applicable to any quic implementation. If we were to switch quic implementations, we would simply need to swap out the error type it was #[from]

File: client/src/quic_client.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    ) -> Result<(), WriteError> {
+        let mut send_stream = connection.connection.open_uni().await?;
+        send_stream.write_all(data).await?;
+        send_stream.finish().await?;
+        Ok(())
+    }
+
+    // Attempts to send data, connecting/reconnecting as necessary
+    // On success, returns the connection used to successfully send the data
+    pub async fn send_buffer(&self, data: &[u8]) -> Result<Arc<NewConnection>, WriteError> {
```
Comment (t-nelson) [inline client/src/quic_client.rs:162]: Similarly here. Avoid exposing implementation details directly in public interfaces

File: client/src/quic_client.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        let connection = {
+            let mut conn_guard = self.connection.lock().await;
+
+            let maybe_conn = (*conn_guard).clone();
+            match maybe_conn {
+                Some(conn) => conn.clone(),
+                None => {
```
Comment (t-nelson) [inline client/src/quic_client.rs:169]: Am I right in the assumption that there is no contention for the mutex in this case?

File: client/src/quic_client.rs (hunk 4 above)
Comment (t-nelson) [reply client/src/quic_client.rs:162]: And throughout the interface

File: client/src/quic_client.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                    *conn_guard = Some(connection.clone());
+                    connection
+                };
+                Self::_send_buffer_using_conn(data, &connection).await?;
+                Ok(connection)
+            }
+        }
+    }
+
+    pub async fn send_batch(&self, buffers: Vec<&[u8]>) -> Result<(), WriteError> {
```
Comment (t-nelson) [inline client/src/quic_client.rs:193]: Seems like this should be returning at least a `Vec<Result<(), WriteError>` so the caller can decide what to do in error cases

File: client/src/quic_client.rs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        // be due to a timed out connection) has succeeded
+
+        if buffers.is_empty() {
+            return Ok(());
+        }
+        let connection = self.send_buffer(buffers[0]).await?;
+
+        let futures = buffers[1..buffers.len()]
+            .par_iter()
+            .chunks(2048)
```
Comment (t-nelson) [inline client/src/quic_client.rs:213]: Can we name and justify the magic number?

File: client/src/thin_client.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply client/src/thin_client.rs:136]: > Removed most of the quic-specific references in thin-client

Did you push the changes? That bool needs to die.  I'd suggest doing it trait-based, as mentioned above for `TpuConnection`

File: client/src/tpu_client.rs [hunk 8]
```diff
@@ -543,6 +544,7 @@ impl LeaderTpuService {
                     slots_in_epoch,
                 ) {
                     Ok(slot_leaders) => {
+                        // TODO(jstarry): Manage list of open connections for upcoming slot leaders
```
Comment (t-nelson) [inline client/src/tpu_client.rs:547]: what does this have to do with QUIC?

File: sdk/src/transport.rs [hunk 9]
```diff
@@ -10,6 +15,12 @@ pub enum TransportError {
     TransactionError(#[from] TransactionError),
     #[error("transport custom error: {0}")]
     Custom(String),
+
+    #[error("transport send error: {0}")]
+    SendError(#[from] WriteError),
+
+    #[error("transport connect error: {0}")]
+    ConnectError(#[from] ConnectError),
```
Comment (t-nelson) [inline sdk/src/transport.rs:23]: Again, we don't want quinn's details here

File: streamer/src/quic.rs [hunk 10]
```diff
@@ -47,7 +47,7 @@ fn configure_server(
         .map_err(|_e| QuicServerError::ConfigureFailed)?;
     let config = Arc::get_mut(&mut server_config.transport).unwrap();
 
-    const MAX_CONCURRENT_UNI_STREAMS: u32 = 1;
+    const MAX_CONCURRENT_UNI_STREAMS: u32 = 4096;
```
Comment (t-nelson) [inline streamer/src/quic.rs:50]: How did we arrive at this number?

File: bench-tps/src/bench.rs [hunk 11]
```diff
@@ -483,10 +484,13 @@ fn do_tx_transfers<T: Client>(
...
                     continue;
                 }
-                client
-                    .async_send_transaction(tx.0)
-                    .expect("async_send_transaction in do_tx_transfers");
+                transactions.push(tx.0);
+            }
+
+            if client.async_send_batch(transactions).is_err() {
+                warn!("send_batch_sync in do_tx_transfers failed");
```
Comment (t-nelson) [inline bench-tps/src/bench.rs:491]:
```suggestion
            if let Err(error) = client.async_send_batch(transactions) {
                warn!("send_batch_sync in do_tx_transfers failed: {}", error);
```

**Review by t-nelson — CHANGES_REQUESTED** [review]
Getting closer!  I'd still like to see this PR broken into three or more, smaller, very specific ones.  It's too large to review properly as is

File: client/src/thin_client.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply client/src/thin_client.rs:136]: So the idea is that when the thin-client is created, the creator chooses whether it's UDP or Quic, and thin-client in turns creates the right type of TpuConnection. I agree that TpuConnection should be a trait, but the bool I think does need to stay in the creation functions for thin-client, or thin-client itself needs to be templated for the TpuConnection type

File: sdk/src/transport.rs (hunk 9 above)
Comment (ryleung-solana) [reply sdk/src/transport.rs:23]: I think the enums are now generic enough that they should apply to any quic implementation (connect error and send error), where basically if the implementation were switched, all we would need to do is replace the error type in the #from. I see something similar done for example at https://www.lpalmieri.com/posts/error-handling-rust/ and actually some of the other errors that were already in there previously do that, for example io::Error, which while admitedly probably more widely used is still in principle the same; if we switched to a new io lib for networking for example, we would need to swap that out. So the idea is that if we were to switch to another quic library we would basically switch for example from SendError(#[from] WriteError), to SendError(#[from] something_else), but the categories SendError and ConnectError should apply to basically any quic implementation

File: client/src/quic_client.rs (hunk 5 above)
Comment (ryleung-solana) [reply client/src/quic_client.rs:169]: Yeah, the idea is that assuming we're only getting one or a few concurrent calls into batch_send at a time, there shouldn't be much lock contention since each batch send only touches the lock once or twice.

File: client/src/thin_client.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply client/src/thin_client.rs]: Have implemented switching between UDP and Quic as a trait; right now it's hardcoded to UDP, will have another PR to allow for switching between the two once all this infrastructure is in.

File: client/src/quic_client.rs (hunk 6 above)
Comment (sakridge) [reply client/src/quic_client.rs]: We can do this in a follow-up I think. The current consumers don't need the specific error.

**Comment by ryleung-solana** [comment]
@t-nelson any other comments?

File: gossip/src/gossip_service.rs [hunk 12]
```diff
@@ -229,7 +238,8 @@ pub fn get_multi_client(
     .unwrap();
     let num_nodes = tpu_addrs.len();
     (
-        ThinClient::new_from_addrs(rpc_addrs, tpu_addrs, transactions_socket),
+        //todo: make it configurable whether to use quic
```
Comment (t-nelson) [inline gossip/src/gossip_service.rs:241]: should just be a matter of parameterizing these functions and choosing at instantiation.  (also moving them out of gossip service?  i never understood why they lived here instead of the thin client module)

typically prefer github issues to `TODO:` as per https://github.com/solana-labs/solana/blob/fb974489a54e32eeff69c8cd4f48459e5910524a/ci/nits.sh#L64

File: client/src/thin_client.rs [hunk 13]
```diff
@@ -142,18 +141,14 @@ impl ThinClient {
...
+        let tpu_connection = C::new(transactions_socket, tpu_addr);
+        Self::new_from_clients(rpc_client, tpu_connection)
     }
 
-    fn new_from_client(
-        tpu_addr: SocketAddr,
-        transactions_socket: UdpSocket,
-        rpc_client: RpcClient,
-    ) -> Self {
+    fn new_from_clients(rpc_client: RpcClient, tpu_connection: C) -> Self {
```
Comment (t-nelson) [inline client/src/thin_client.rs:148]:
```suggestion
    fn new_from_client(rpc_client: RpcClient, tpu_connection: C) -> Self {
```

right?  it doesn't take multiple clients as input

File: client/src/thin_client.rs [hunk 14]
```diff
@@ -142,18 +141,14 @@ impl ThinClient {
         timeout: Duration,
     ) -> Self {
         let rpc_client = RpcClient::new_socket_with_timeout(rpc_addr, timeout);
-        Self::new_from_client(tpu_addr, transactions_socket, rpc_client)
+        let tpu_connection = C::new(transactions_socket, tpu_addr);
+        Self::new_from_clients(rpc_client, tpu_connection)
```
Comment (t-nelson) [inline client/src/thin_client.rs:145]:
```suggestion
        Self::new_from_client(rpc_client, tpu_connection)
```

File: client/src/thin_client.rs [hunk 15]
```diff
@@ -117,22 +118,20 @@ impl ClientOptimizer {
...
+    /// (currently hardcoded to UDP)
     pub fn new(rpc_addr: SocketAddr, tpu_addr: SocketAddr, transactions_socket: UdpSocket) -> Self {
-        Self::new_from_client(
-            tpu_addr,
-            transactions_socket,
-            RpcClient::new_socket(rpc_addr),
-        )
+        let tpu_connection = C::new(transactions_socket, tpu_addr);
+
+        Self::new_from_clients(RpcClient::new_socket(rpc_addr), tpu_connection)
```
Comment (t-nelson) [inline client/src/thin_client.rs:134]:
```suggestion
        Self::new_from_client(RpcClient::new_socket(rpc_addr), tpu_connection)
```

File: client/src/quic_client.rs [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+        Self {
+            runtime,
+            endpoint,
+            connection: Arc::new(Mutex::new(None)),
+            addr,
+        }
+    }
+
+    // If this function becomes public, it should be changed to
+    // not expose details of the specific Quic implementation we're using
```
Comment (t-nelson) [inline client/src/quic_client.rs:120]: :heart:

File: client/src/quic_client.rs [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+            .chunks(MAX_CONCURRENT_STREAMS)
+            .into_iter()
+            .map(|buffs| {
+                join_all(
+                    buffs
+                        .into_iter()
+                        .map(|buf| Self::_send_buffer_using_conn(&buf[..], &connection)),
+                )
+            })
+            .collect::<Vec<_>>();
```
Comment (t-nelson) [inline client/src/quic_client.rs:201]: do we need the collect?  usually clippy complains when they're unnecessary, but seems like we can just loop over the iterator directly below

**Review by t-nelson — APPROVED** [review]
lgtm now! r+ nits

File: gossip/src/gossip_service.rs (hunk 12 above)
Comment (ryleung-solana) [reply gossip/src/gossip_service.rs]: Changed the todo to TODO. I think this can be done in a followup - I was planning a followup PR to allow to choose the transport protocol via a command line arg.

Comment (t-nelson) [reply gossip/src/gossip_service.rs]: ah I was suggesting to just delete the todo-comment and open an issue on github in its place, but I see that I referenced the wrong comment in `nits.sh` :sweat_smile: 

we still let carl add them, so one more won't hurt. for future reference, though!
